File: src/logger.ts
/**
 * Logger interface for BiDi client
 * Allows consumers to provide their own logging implementation
 */

export interface Logger {
  log(message: string): void;
  debug(message: string): void;
  error(message: string, error?: unknown): void;
}

/**
 * Default no-op logger
 */
class NoOpLogger implements Logger {
  log(_message: string): void {}
  debug(_message: string): void {}
  error(_message: string, _error?: unknown): void {}
}

/**
 * Simple console logger
 */
export class ConsoleLogger implements Logger {
  constructor(private debug_enabled = false) {}

  log(message: string): void {
    console.log(message);
  }

  debug(message: string): void {
    if (this.debug_enabled) {
      console.log(`[DEBUG] ${message}`);
    }
  }

  error(message: string, error?: unknown): void {
    console.error(message, error);
  }
}

let currentLogger: Logger = new NoOpLogger();

export function setLogger(logger: Logger): void {
  currentLogger = logger;
}

export function getLogger(): Logger {
  return currentLogger;
}

export function log(message: string): void {
  currentLogger.log(message);
}

export function logDebug(message: string): void {
  currentLogger.debug(message);
}

export function logError(message: string, error?: unknown): void {
  currentLogger.error(message, error);
}

File: python/firefox_bidi_client/logger.py
import sys


class Logger:
    def log(self, message: str) -> None: ...
    def debug(self, message: str) -> None: ...
    def error(self, message: str, error=None) -> None: ...


class NoOpLogger(Logger):
    def log(self, message: str) -> None: pass
    def debug(self, message: str) -> None: pass
    def error(self, message: str, error=None) -> None: pass


class ConsoleLogger(Logger):
    def __init__(self, debug_enabled: bool = False):
        self._debug_enabled = debug_enabled

    def log(self, message: str) -> None:
        print(message)

    def debug(self, message: str) -> None:
        if self._debug_enabled:
            print(f"[DEBUG] {message}")

    def error(self, message: str, error=None) -> None:
        print(message, error, file=sys.stderr)


_current_logger: Logger = NoOpLogger()


def set_logger(logger: Logger) -> None:
    global _current_logger
    _current_logger = logger


def get_logger() -> Logger:
    return _current_logger


def log(message: str) -> None:
    _current_logger.log(message)


def log_debug(message: str) -> None:
    _current_logger.debug(message)


def log_error(message: str, error=None) -> None:
    _current_logger.error(message, error)
